smart-contracts/assembly/contracts/NFT/NFT-internals.ts

```diff
@@ -1,282 +1,231 @@
-  Storage.set(key, '');
+export function _ownerOf(tokenId: u256): string {
+  const key = ownerKey(tokenId);
+  if (Storage.has(key)) {
+    return bytesToString(Storage.get(key));
+  }
+  return '';
 }
 
-export function _getApproved(tokenId: u64): string {
-  const key = approvedTokenKey + tokenId.toString();
-  if (!Storage.has(key)) return '';
-
-  return Storage.get(key);
+/**
+ * Returns the name of the contract.
+ */
+export function _name(): string {
+  if (Storage.has(NAME_KEY)) {
+    return bytesToString(Storage.get(NAME_KEY));
+  }
+  return '';
 }
 
-export function _isApproved(address: string, tokenId: u64): bool {
-  if (address.length === 0) return false;
-  const approvedAddress = _getApproved(tokenId);
-  return approvedAddress === address;
+/**
+ * Returns the symbol of the contract.
+ */
+export function _symbol(): string {
+  if (Storage.has(SYMBOL_KEY)) {
+    return bytesToString(Storage.get(SYMBOL_KEY));
+  }
+  return '';
 }
 
-export function _setApprovalForAll(
-  owner: string,
-  operator: string,
-  approved: bool,
-): void {
-  assert(owner != operator, 'Not allowed to set approval for all for yourself');
-
-  const key = approvedForAllTokenKey + owner + operator;
-
-  Storage.set(key, approved.toString());
+/**
+ * Change or reaffirm the approved address for an NFT
+ *
+ * @param tokenId - The NFT to approve
+ * @param approved - The new approved NFT controller
+ */
+export function _approve(approved: string, tokenId: u256): void {
+  assert(_isAuthorized(Context.caller().toString(), tokenId), 'Unauthorized');
+  const key = allowanceKey(tokenId);
+  Storage.set(key, stringToBytes(approved));
 }
 
-export function _isApprovedForAll(owner: string, operator: string): bool {
-  const key = approvedForAllTokenKey + owner + operator;
-
-  if (!Storage.has(key)) return false;
-
-  const approved = Storage.get(key);
-
-  return approved == 'true';
+/**
+ * Get the approved address for a single NFT
+ * @param tokenId - Id of the NFT
+ * @returns Address of the approved owner of the NFT
+ */
+export function _getApproved(tokenId: u256): string {
+  const key = allowanceKey(tokenId);
+  if (Storage.has(key)) {
+    return bytesToString(Storage.get(key));
+  }
+  return '';
 }
 
-// ==================================================== //
-// ====             General Assertions             ==== //
-// ==================================================== //
-
-export function assertIsMinted(tokenId: u64): void {
-  assert(_onlyMinted(tokenId), `Token ${tokenId.toString()} is not minted`);
+/**
+ * Returns the allowance set on the owner's account for the spender.
+ *
+ * @param operator - address of the operator
+ * @param tokenId - tokenId of the token
+ * @returns
+ */
+export function _isApproved(operator: string, tokenId: u256): bool {
+  const allowKey = allowanceKey(tokenId);
+  if (Storage.has(allowKey)) {
+    return bytesToString(Storage.get(allowKey)) == operator;
+  }
```

Comment:
if i understand well , only one allowed operator is possible ?
maybe it would be more efficient to store the operator address in the key, then this function would become
```
const allowKey = allowanceKey(tokenId, operator);
return Storage.has(allowKey);

Reply:
Good point. But this imply that we should delete allowanceKey when revoking allowance

Reply:
After all, I'm not sure it's a good idea anymore. See, if the operator address is in the key, then when calling 'approve' method, you'll have to check in the storage if an approval is not already set on this id. We would have to filter the storage based on the prefix+tokenId. The same would also apply for the _getApproved method.
I think we should keep it simple. Your suggestion implies heavy modification and complex storage access for little gas efficiency. 

But I'll keep the delete allowanceKey when approval is called with zero address as parameter